src/indexer.c

```diff
@@ -240,248 +240,174 @@ int IsKeyIndexed(RedisModuleString *ts_key) {
+            if (!doesNotExist) {
+                found = true;
             }
         }
+        free(dicts);
 
-        // switch between the minimal predicate and the predicate in the first place
-        if (minIndex != 0) {
-            QueryPredicate temp = index_predicate[minIndex];
-            index_predicate[minIndex] = index_predicate[0];
-            index_predicate[0] = temp;
+        if (inclusion != found) {
+            return false;
         }
     }
+
+    return true;
 }
 
 RedisModuleDict *QueryIndex(RedisModuleCtx *ctx,
                             QueryPredicate *index_predicate,
                             size_t predicate_count) {
-    RedisModuleDict *result = NULL;
-
     PromoteSmallestPredicateToFront(ctx, index_predicate, predicate_count);
 
-    // EQ or Contains
-    for (int i = 0; i < predicate_count; i++) {
-        if (index_predicate[i].type == EQ || index_predicate[i].type == CONTAINS ||
-            index_predicate[i].type == LIST_MATCH) {
-            result = QueryIndexPredicate(ctx, &index_predicate[i], result);
-            if (result == NULL) {
-                return RedisModule_CreateDict(ctx);
-            }
-        }
+    RedisModuleDict *res = RedisModule_CreateDict(ctx);
+    QueryPredicate *predicate = &index_predicate[0];
+
+    if (!IS_INCLUSION(predicate->type)) {
+        return res;
     }
 
-    // The next two types of queries are reducers so we run them after the matcher
-    // NCONTAINS or NEQ
-    for (int i = 0; i < predicate_count; i++) {
-        if (index_predicate[i].type == NCONTAINS || index_predicate[i].type == NEQ ||
-            index_predicate[i].type == LIST_NOTMATCH) {
-            result = QueryIndexPredicate(ctx, &index_predicate[i], result);
-            if (result == NULL) {
-                return RedisModule_CreateDict(ctx);
+    RedisModuleDict **dicts = NULL;
+    size_t dicts_size;
+    GetPredicateKeysDicts(ctx, &index_predicate[0], &dicts, &dicts_size);
+
+    for (size_t i = 0; i < dicts_size; i++) {
+        RedisModuleDict *dict = dicts[i];
+        if (dict == NULL) {
+            continue;
+        }
+
+        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(dict, "^", NULL, 0);
+        char *currentKey;
+        size_t currentKeyLen;
+        while ((currentKey = RedisModule_DictNextC(iter, &currentKeyLen, NULL)) != NULL) {
+            if (_isKeySatisfyAllPredicates(
+                    ctx, currentKey, currentKeyLen, index_predicate, predicate_count)) {
+                RedisModule_DictSetC(res, currentKey, currentKeyLen, (void *)1);
             }
         }
+        RedisModule_DictIteratorStop(iter);
     }
 
-    if (result == NULL) {
-        return RedisModule_CreateDict(ctx);
-    }
+    free(dicts);
 
     if (unlikely(isTrimming)) {
-        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(result, "^", NULL, 0);
+        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(res, "^", NULL, 0);
```

Comment:
Looks like we don't have tests to cover this, is it covered in other places?

Reply:
@danni-m I think we this covered in the Automation tests.